transaction/core/src/tx.rs

```diff
@@ -333,56 +334,74 @@ impl TxOut {
     /// This uses a defaulted (all zeroes) MemoPayload.
     ///
     /// # Arguments
+    /// * `block_version` - Structural rules to target
     /// * `amount` - Amount contained within the TxOut
     /// * `recipient` - Recipient's address.
     /// * `tx_private_key` - The transaction's private key
     /// * `hint` - Encrypted Fog hint for this output.
     pub fn new(
+        block_version: BlockVersion,
         amount: Amount,
         recipient: &PublicAddress,
         tx_private_key: &RistrettoPrivate,
         hint: EncryptedFogHint,
-    ) -> Result<Self, AmountError> {
-        TxOut::new_with_memo(amount, recipient, tx_private_key, hint, |_| {
-            Ok(Some(MemoPayload::default()))
-        })
-        .map_err(|err| match err {
-            NewTxError::Amount(err) => err,
-            // Memo error is unreachable because the memo_fn we passed is infallible
-            NewTxError::Memo(_) => unreachable!(),
-        })
+    ) -> Result<Self, NewTxError> {
+        TxOut::new_with_memo(
+            block_version,
+            amount,
+            recipient,
+            tx_private_key,
+            hint,
+            |_| Ok(MemoPayload::default()),
+        )
     }
 
     /// Creates a TxOut that sends `value` to `recipient`, with a custom memo
     /// attached. The memo is produced by a callback function which is
     /// passed the value and tx_public_key.
     ///
     /// # Arguments
+    /// * `block_version` - Structural rules to target
     /// * `amount` - Amount contained within the TxOut
     /// * `recipient` - Recipient's address.
     /// * `tx_private_key` - The transaction's private key
     /// * `hint` - Encrypted Fog hint.
     /// * `memo_fn` - A callback taking MemoContext, which produces a
     ///   MemoPayload, or a NewMemo error
     pub fn new_with_memo(
+        block_version: BlockVersion,
         amount: Amount,
         recipient: &PublicAddress,
         tx_private_key: &RistrettoPrivate,
         hint: EncryptedFogHint,
-        memo_fn: impl FnOnce(MemoContext) -> Result<Option<MemoPayload>, NewMemoError>,
+        memo_fn: impl FnOnce(MemoContext) -> Result<MemoPayload, NewMemoError>,
     ) -> Result<Self, NewTxError> {
         let target_key = create_tx_out_target_key(tx_private_key, recipient).into();
         let public_key = create_tx_out_public_key(tx_private_key, recipient.spend_public_key());
 
         let shared_secret = create_shared_secret(recipient.view_public_key(), tx_private_key);
 
-        let masked_amount = MaskedAmount::new(amount, &shared_secret)?;
+        let mut masked_amount = MaskedAmount::new(amount, &shared_secret)?;
 
         let memo_ctxt = MemoContext {
```

Comment:
Consider moving this declaration into the `if block_version.e_memo_feature_is_supported()` block below.